Slava

Arista213 opened the pull request:
Task 1.
При старте диалога, чат-бот рассказывает, что он умеет и как с ним взаимодействовать. Затем в бесконечном цикле задает вопрос, получает от пользователя ответ и оценивает его правильность, задаёт следующий вопрос и так далее. В любой момент можно командой /help попросить бота ещё раз рассказать о себе, и пояснить как с ним взаимодействовать.

Код реализующий хранение списка блюд находится в классе DishService - dishes.
Код реализующий работу с консолью находится в классе ConsoleService, который реализует интерфейс ICommunicationService созданный, чтобы было вводить в выводить данные из бота. Логика вся разделена.
Все команды покрыты тестами.

Тесты для проверки функциональности находятся в BotTests.
Тесты для режима администратора находятся в AdminModeCommandsTest.
Тесты для вывод рецептов по продуктам/названию блюда находятся в DishesTests

Arista213 committed:
Console application

Arista213 committed:
Add dish class

Arista213 committed:
Add JSONWorker, saving and loading json files

Arista213 committed:
Add threading

Arista213 committed:
Add threading

Arista213 committed:
Merge pull request #2 from Arista213/NewCommands

New commands

Arista213 committed:
Add recipe by ingredients and recipe by name

Arista213 committed:
Merge branch 'NewCommands' into Slava

Arista213 committed:
Delete BotConfiguration.java

wintrnite committed:
Add admin mode

Arista213 committed:
Merge branch 'vasya' into Slava

wintrnite committed:
Added tests

wintrnite committed:
Fixed test

wintrnite committed:
Added all tests

Arista213 committed:
Delete commands hashset

Arista213 committed:
Add help command

wintrnite committed:
Add tests for admin mode

Arista213 committed:
Add dishes tests

Arista213 committed:
refactor

delete threading
add communicate service
add recipe class
add message class
rewrite tests

eez-eh commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,7 @@
+import logic.communication.ConsoleService;
+
+public class Main {
```
Прогнать все файлы с Java кодом через форматтер (Ctrl + Alt + L):
https://devcolibri.com/%D0%B3%D0%BE%D1%80%D1%8F%D1%87%D0%B8%D0%B5-%D0%BA%D0%BB%D0%B0%D0%B2%D0%B8%D1%88%D0%B8-intellij-idea/

eez-eh commented on README.md:
```diff
@@ -1 +1,12 @@
+Бот способен выдать название блюда и его рецепт по имеющимся ингредиентам или рецепт блюда по его названию.
+
+# Комманды
++ /start - Начало работы с ботом.
```
API обсудим устно.

eez-eh commented on src/main/java/logic/communication/ICommunicateService.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Сервис, ответственный за принятие сообщений пользователя и отправления ему ответов.
+ */
+public interface ICommunicateService {
```
MessageProvider

eez-eh commented on src/main/java/logic/Bot.java:
```diff
@@ -0,0 +1,51 @@
+            default -> new UnknownCommand();
+        };
+
+        command.process(this);
```
```
public void runCommand(Message message, User user) {
    ICommand command = commandMap.get(message.getText().toLowerCase());
    command.execute(message, user);
  }
```

eez-eh commented on src/main/java/logic/Bot.java:
```diff
@@ -0,0 +1,51 @@
+     */
+    public void runCommand(String input) {
+        ICommand command = switch (input) {
+            case "/start" -> new Start();
```
```
public enum AppCommands {
    START, // Добавить что-то
    HELP,// Показать помощь

```

eez-eh commented on src/main/java/logic/commands/RemoveRecipeByAdmin.java:
```diff
@@ -0,0 +1,23 @@
+            return;
+        }
+
+        bot.setOutput("Введите название блюда, которое вы хотите удалить");
```
```
/**
 * Константы бота
 */
public enum AppConstants {
    BOT_NAME ("......"),// Имя бота
    /**
     * Конструкция присвоения enum констант с их значениями
     */
    private final String value;
    AppConstants(String value){
        this.value = value;
    }
    public String toStringValue(){
        return value;
    }
}
```

eez-eh reviewed (commented):
![image](https://user-images.githubusercontent.com/28267530/138096996-89d50884-aa05-4833-8e70-6a3f2eab3c7f.png)


Примерная структура

Arista213 committed:
refactor

add Product
add ProductService
add ProductServiceTests
add AppConstants
add abstract class Command which implements ICommand interface
rename DishService -> DishApi
rename name -> title
rework ICommunicateService -> abstract class MessageProvider
rework all commands, now they extends abstract class command.
delete BotTests

Arista213 committed:
refactor

rework getValidDishes
rework findDishByTitle
rework isProductsFit

eez-eh commented on src/main/java/api/DishApi.java:
```diff
@@ -0,0 +1,83 @@
+    /**
+     * Здесь хранятся все блюда, с которыми можно взаимодействовать.
+     */
+    private static List<Dish> dishes;
```
Давайте ещё сделаем метод getAll и будем возвращать все dishes

eez-eh commented on src/main/java/api/DishApi.java:
```diff
@@ -0,0 +1,83 @@
+     * @return блюда, которые могут быть приготовлены.
+     */
+    public static Set<Dish> getValidDishes(Set<Product> products) {
+        HashSet<Dish> validDishes = new HashSet<>();
```
Set<Dish> слева (стараться тут и везде в других местах работать с интерфейсными типами для обеспечения полиморфного кода)

eez-eh commented on src/main/java/api/DishApi.java:
```diff
@@ -0,0 +1,83 @@
+     * @param products имеющиеся у пользователя.
+     * @return блюда, которые могут быть приготовлены.
+     */
+    public static Set<Dish> getValidDishes(Set<Product> products) {
```
`getAvailableForUser`

eez-eh commented on src/main/java/api/DishApi.java:
```diff
@@ -0,0 +1,83 @@
+    /**
+     * Проверка подходит ли список продуктов рецепту блюда.
+     */
+    public static boolean isProductsFit(Dish dish, Set<Product> products) {
```
private

eez-eh commented on src/main/java/api/DishApi.java:
```diff
@@ -0,0 +1,83 @@
+     * Удалить блюдо из "базы блюд".
+     */
+    public static void remove(String dishTitle) {
+        for (int i = 0; i < dishes.size(); i++) {
```
Цикл for-each. Для списков метод remove может удалять сразу объект вашего класса, если его найдёте вначале (если переопределены equals и hashCode). Аналогично для других методов.

eez-eh commented on src/main/java/api/DishApi.java:
```diff
@@ -0,0 +1,83 @@
+     * @return блюдо.
+     */
+    public static Dish findDishByTitle(String title) {
+        Optional<Dish> dish = dishes.stream().filter(x -> Objects.equals(x.title, title)).findFirst();
```
`return dishes.stream().filter(x -> Objects.equals(x.title, title)).findFirst().orElse(null);`

eez-eh commented on src/main/java/model/User.java:
```diff
@@ -0,0 +1,22 @@
+ * Сущность пользователя.
+ */
+public class User {
+    private boolean m_isAdmin;
```
`isAdmin`

eez-eh commented on src/main/java/message/MessageProvider.java:
```diff
@@ -0,0 +1,13 @@
+ * Абстракция, ответственная за принятие сообщений пользователя и отправления ему ответов.
+ */
+public abstract class MessageProvider {
+    public Message getMessage() {
```
abstract

eez-eh commented on src/main/java/message/MessageProvider.java:
```diff
@@ -0,0 +1,13 @@
+        return null;
+    }
+
+    public void sendMessage(Message message) {
```
abstract

eez-eh commented on src/main/java/model/Dish.java:
```diff
@@ -0,0 +1,46 @@
+    private final Recipe recipe;
+
+    public Dish() {
+        title = null;
```
Избыточные присвоения в конструкторе, у полей, объявленных выше и так есть значения по умолчанию.

eez-eh commented on .gitignore:
```diff
@@ -1,3 +1,60 @@
+# Maven
+log/
+target/
```
target/*

eez-eh reviewed (commented):
toFix

Arista213 committed:
refactor

HashSet to Set
rename getValidDishes -> getAvailableForUser
public isProductFit -> private isProductFit
